### DISCUSSION TITLE
[Feature] Support for AWS Lambda

### DISCUSSION THREAD

**Discussion by jchnkl** [immich-app/immich#10394] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I've been looking at AWS Lambda for running Immich. From what I can see it should be possible and (imho) make sense too, but this is also a call for more experienced opinions.

Lambda can run of docker images, either from their pre-built runtime enabled python / node / etc offerings or by including the Lambda runtime. There are two main images in immich, immich-server (based on Node) and immich-machine-learning (based on Python). Both would "merely" need to include the runtime as described [here](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-image.html) and [here](https://docs.aws.amazon.com/lambda/latest/dg/python-image.html). Alternatively the images would need to be rebuilt using AWS' base image.

Including the runtime seems to be the simplest solution, all it requires is installing the `aws-lambda-ric` package (`RUN npm install aws-lambda-ric`) and changing the `ENTRYPOINT` and `CMD`:

```
# Set runtime interface client as default command for the container runtime
ENTRYPOINT ["/usr/local/bin/npx", "aws-lambda-ric"]
# Pass the name of the function handler as an argument to the runtime
CMD ["index.handler"]
```

However I'm unsure what to do with the original `ENTRYPOINT`s and `CMD`s:

```
# server/Dockerfile for example
ENTRYPOINT ["tini", "--", "/bin/bash"]
CMD ["start.sh"]
```

Running Immich on Lambda (with S3 Glacier Instant) would be a very low-cost, yet always available solution for low-volume users who don't need the service for days (or even weeks) but who want access when not at home.

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
This is a neat idea, but I don't really think this makes sense for our target audience & we're probably not interested in supporting it upstream. If this is indeed as simple as you say then it should be pretty straightforward to tack on support in a fork though.

**Reply by jchnkl** [reply]
Well, my intention was to also get an opinion if it's really as simple as it looks like or if there are some pitfalls that I don't see. :) Given that one has to change the entrypoint I'm not sure if that could really work for Immich.

**Reply by bo0tzz** [reply]
Immich uses workers for background jobs via a redis queue. That doesn't seem to me like it really fits onto the lambda model, and you'd probably run into some issues around it.

**Reply by jchnkl** [reply]
Thanks for your assessment! Another option might be Fargate which can run arbitrary containers without the lambda runtime and is not limited to 15 minutes. However, this might get expensive quickly, even for very little use (<1h / day on average).
Alternatively one might spin up an EC2 instance when required, but this adds several minutes of latency. The cheapest (and least wasteful) option could be a suspended server at home accessed through VPN, also I'd favor a cloud solution so I can turn off electronics at home when I'm traveling.
So, my use case is to have a self-hosted Immich instance available on demand without having to run a server 24/7.
Just out of curiosity, what's Immich's target audience?

**Reply by bo0tzz** [reply]
A server at home or a dedi is likely the most effective option, doing stuff like auto suspend/spinup is probably doable but likely gets complex pretty quickly. If you did investigate the lambda/EC2 direction then I'd probably recommend trying to cram the api worker into a lambda, and then spinning up EC2 for the background jobs (but you'd still need postgres and redis always running, so maybe that defeats the point a bit).

> what's Immich's target audience?

We haven't specifically formulated a target audience, but "self-hosters" probably covers most of it :P The vast majority of our users are running on a machine at home, with some on dedis or cloud VMs.

**Reply by jchnkl** [reply]
Thanks! I guess I'm pretty much the target audience :)
My only additional "requirement" would be to have the service available on request only because I'd use it rarely and I think it'd be a waste to have something running 24/7. 
I also wouldn't mind to have immich running on my phone pulling in the data from a cloud service (eg S3). But as far as I can see docker on android is not really working atm. 
So far my preferred solution would be to have a lambda instance which gets started on a http request which then handles the authentication (for which one would need postgres, but maybe that could be fitted to an sqlite or even a json file with hashes that gets synced with the dB). On successful login the instance would spin up the server container which in turn would spin up the machine learning container if necessary. When all users are logged out the whole thing could be spun down again, leaving only persistent data on elastic volumes or S3.
That'd require some adaptions and I think I don't have the resources to keep up with immich's development pace :)
Maybe I take the easier route and build a static site with some basic auth and a button to launch and stop an ec2 instance on request. Not as user friendly but probably good enough for me.

**Reply by bo0tzz** [reply]
I think fly.io makes it pretty easy to do something like this, including the option to have postgres scale to zero as well. That's the route I'd recommend attempting. That said, with the prices that cloud services charge, I wouldn't be surprised if it's cheaper to just have a box at home that's on 24/7.

**Reply by jchnkl** [reply]
Thanks for the hint to fly.io! I definitively will try to get immich running there for one or two months to see what the cost could be.
You're right, that a at home box could be cheaper depending on the cost of electricity. But when you start to factor in hardware cost (for 5 years? for 10 years? 2 or 4TB hdd?), it really gets narrow and in some scenarios it's not cheaper (trust me, I created a big spreadsheet just for this :D). Apart from that you have to worry about maintenance, either by replacing hardware or by updating software (other than immich).
Another important point: When I'm traveling I like to unplug all of my electronics at home. It gives me a lot of peace of mind to know that this stuff won't burn down my home.
